conformance/tests/generics_basic.py

```diff
@@ -0,0 +1,154 @@
+BadConstraint2 = TypeVar('BadConstraint', str, T)  # Type error
+
+# > Subtypes of types constrained by a type variable should be treated
+# > as their respective explicitly listed base types in the context of the
+# > type variable.
+
+class MyStr(str): ...
+
+def test_concat_subtype(s: str, b: bytes, a: Any, m: MyStr) -> None:
+    assert_type(concat(m, m), str)
+    assert_type(concat(m, s), str)
+    concat(m, b)  # Type error
+
+    # TODO: should these be str or Any?
+    # reveal_type(concat(m, a))
+    # reveal_type(concat(a, m))
+
+# Specification: https://typing.readthedocs.io/en/latest/spec/generics.html#user-defined-generic-classes
+
+# > You can include a ``Generic`` base class to define a user-defined class
+# > as generic.
+
+from logging import Logger
+from collections.abc import Iterable
+
+class LoggedVar(Generic[T]):
+    def __init__(self, value: T, name: str, logger: Logger) -> None:
+        self.name = name
+        self.logger = logger
+        self.value = value
+
+    def set(self, new: T) -> None:
+        self.log('Set ' + repr(self.value))
+        self.value = new
+
+    def get(self) -> T:
+        self.log('Get ' + repr(self.value))
+        return self.value
+
+    def log(self, message: str) -> None:
+        self.logger.info('{}: {}'.format(self.name, message))
+
+
+def zero_all_vars(vars: Iterable[LoggedVar[int]]) -> None:
+    for var in vars:
+        var.set(0)
+        assert_type(var.get(), int)
+
+
+# > A generic type can have any number of type variables, and type variables
+# > may be constrained.
+
+S = TypeVar('S')
+
+class Pair1(Generic[T, S]):
+    ...
+
+# > Each type variable argument to ``Generic`` must be distinct.
+
+class Pair2(Generic[T, T]):   # Type error
+      ...
+
+# > The ``Generic[T]`` base class is redundant in simple cases where you
+# > subclass some other generic class and specify type variables for its
+# > parameters.
+
+from collections.abc import Iterator
+
+class MyIter1(Iterator[T]):
+    ...
+
+class MyIter2(Iterator[T], Generic[T]):
+    ...
+
+def test_my_iter(m1: MyIter1[int], m2: MyIter2[int]):
+    assert_type(next(m1), int)
+    assert_type(next(m2), int)
+
+# > You can use multiple inheritance with ``Generic``
+
+from collections.abc import Sized, Container
+
+class LinkedList(Sized, Generic[T]):
+    ...
+
+K = TypeVar('K')
+V = TypeVar('V')
+
+class MyMapping(Iterable[tuple[K, V]], Container[tuple[K, V]], Generic[K, V]):
+    ...
+
+# > Subclassing a generic class without specifying type parameters assumes
+# > ``Any`` for each position.  In the following example, ``MyIterable``
+# > is not generic but implicitly inherits from ``Iterable[Any]``::
+
+class MyIterableAny(Iterable):  # Same as Iterable[Any]
+    ...
+
+def test_my_iterable_any(m: MyIterableAny):
+    assert_type(iter(m), Iterator[Any])
+
+# > Generic metaclasses are not supported
+
+class GenericMeta(type, Generic[T]): ...
+
+class GenericMetaInstance(metaclass=GenericMeta[T]):  # Type error
+    ...
```

Comment:
It may also be useful to add tests for accessing a TypeVar as a runtime object — to make sure that this isn't disallowed.

```python
assert_type(T.__name__, str)
assert_type(T.__bound__, Any | None)
T.other # Type error

runtime_typevar: TypeVar = T # OK
```